### PR TITLE
Test that documentation does not contain any warnings.

### LINKED ISSUE #3944: Update the documentation to not generate any compiler warning.
We can utilize `test/cmdlineTests.sh` to find warnings in the documentation.

After this, we should also change this script to force no warnings.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/3944

 - [x] fix all errors
 - [x] check that all code examples actually start with `pragma` - should we have a separate test about that?

### CODE REVIEW

**Comment by axic** [comment]
The script change looks good.

**Comment by axic** [comment]
> check that all code examples actually start with pragma - should we have a separate test about that?

Yes, sounds good.

File: docs/contracts.rst [hunk 1]
```diff
@@ -616,14 +616,13 @@ Like any function, the fallback function can execute complex operations as long
     }
```
Comment (axic) [inline docs/contracts.rst:614]: Not sue what is this `Sink` contract useful for, judging from the context.

**Comment by chriseth** [comment]
Should be all good now.

**Comment by chriseth** [comment]
This now extracts way more code snippets - basically everything that is indented with spaces and has a line that starts with `pragma `, `library `, `contract ` or `interface `. The extraction of course has some false positives, so I had to change some line breaks in the text.

This now especially also matches some snippets in the style guide which is a good side-effect, I would say.

Ready for review.

File: docs/contracts.rst [hunk 2]
```diff
@@ -657,9 +657,9 @@ externally visible functions differ by their Solidity types but not by their ext
 
 ::
 
-    // This will not compile
     pragma solidity ^0.4.16;
 
+    // This will not compile
```
Comment (axic) [inline docs/contracts.rst:662]: This has to be above the pragma to disable compilation of it.

File: docs/contracts.rst [hunk 3]
```diff
@@ -1080,12 +1082,11 @@ error "Linearization of inheritance graph impossible".
 
 ::
 
-    // This will not compile
```
Comment (axic) [inline docs/contracts.rst:1082]: This has to be the first line to disable compilation of it.

File: docs/contracts.rst [hunk 4]
```diff
@@ -1080,12 +1082,11 @@ error "Linearization of inheritance graph impossible".
...
-
     pragma solidity ^0.4.0;
 
     contract X {}
     contract A is X {}
+    // This will not compile
```
Comment (axic) [inline docs/contracts.rst:1089]: I would suggest leave a comment here: "This is the reason this won't compile"

File: docs/control-structures.rst [hunk 5]
```diff
@@ -287,12 +292,13 @@ These can then either be assigned to newly declared variables or to pre-existing
         }
 
         function g() public {
-            // Variables declared with type and assigned from the returned tuple.
-            (uint x, bool b, uint y) = f();
+            // Variables declared with type and assigned from the returned tuple,
+            // not all elements have to be specified (but the amount must match).
```
Comment (axic) [inline docs/control-structures.rst]: Amount? Maybe count? Number of them?

File: docs/control-structures.rst [hunk 6]
```diff
@@ -372,9 +381,8 @@ In any case, you will get a warning about the outer variable being shadowed.
 
  ::
 
-    // This will not compile
-
     pragma solidity >0.4.24;
+    // This will not compile
```
Comment (axic) [inline docs/control-structures.rst:385]: Again, cannot move this.

File: docs/frequently-asked-questions.rst [hunk 7]
```diff
@@ -372,15 +374,14 @@ contract level) with ``arrayname.length = <some new length>;``. If you get the
 
 ::
 
-    // This will not compile
-
     pragma solidity ^0.4.18;
 
+    // This will not compile
```
Comment (axic) [inline docs/frequently-asked-questions.rst:379]: Same.

File: docs/types.rst [hunk 8]
```diff
@@ -646,8 +647,9 @@ Allocating Memory Arrays
 ^^^^^^^^^^^^^^^^^^^^^^^^
 
 Creating arrays with variable length in memory can be done using the ``new`` keyword.
-As opposed to storage arrays, it is **not** possible to resize memory arrays by assigning to
-the ``.length`` member.
+As opposed to storage arrays, it is **not** possible to resize memory arrays (e.g. by assigning to
+the ``.length`` member). You either have to calculate the required size in advance
+or crete a new memory array and copy every element.
```
Comment (axic) [inline docs/types.rst]: crete -> create

File: docs/types.rst [hunk 9]
```diff
@@ -693,10 +696,9 @@ possible:
 
 ::
 
-    // This will not compile.
-
     pragma solidity ^0.4.0;
 
+    // This will not compile.
```
Comment (axic) [inline docs/types.rst:714]: Cannot be moved.

File: test/cmdlineTests.sh [hunk 10]
```diff
@@ -169,8 +177,25 @@ TMPDIR=$(mktemp -d)
...
+        if grep -E 'DeclarationError:|// ----' "$f" >/dev/null
+        then
+            continue
+        fi
         echo "$f"
-        compileFull "$TMPDIR/$f"
+        opts=''
+        # We expect errors if explicitly stated, or if imports
+        # are used (in the style guide)
+        if grep -E "This will not compile|// DeclarationError:|import \"" "$f" >/dev/null
```
Comment (chriseth) [inline test/cmdlineTests.sh]: Should remove `DeclarationError`

File: docs/contracts.rst (hunk 2 above)
Comment (chriseth) [reply docs/contracts.rst:662]: I changed the script to accept it anywhere: https://github.com/ethereum/solidity/pull/4788/files#diff-a8a5760144e03e410389fb0afd20e8c8R190

File: docs/contracts.rst (hunk 4 above)
Comment (chriseth) [reply docs/contracts.rst:1089]: do you suggest to use a different trigger phrase? If I change this line, the tests will break.

**Comment by chriseth** [comment]
@axic addressed everything.

**Comment by chriseth** [comment]
Relevant tests are green.

File: test/cmdlineTests.sh [hunk 11]
```diff
@@ -43,18 +43,42 @@ function printError() { echo "$(tput setaf 1)$1$(tput sgr0)"; }
...
-    if [ $failed -ne 0 ]
+    if [[ \
+        "$exit_code" -ne "$expected_exit_code" || \
+            ( $expect_output -eq 0 && -n "$errors" ) || \
+            ( $expect_output -ne 0 && -z "$errors" ) \
+    ]]
     then
-        printError "Compilation failed on:"
-        echo "$output"
+        printError "Unexpected compilaciton result:"
```
Comment (ekpyron) [inline test/cmdlineTests.sh]: Typo ``compilaciton``.

File: test/cmdlineTests.sh [hunk 12]
```diff
@@ -173,8 +181,25 @@ TMPDIR=$(mktemp -d)
     "$REPO_ROOT"/scripts/isolate_tests.py "$REPO_ROOT"/docs/ docs
     for f in *.sol
     do
+        # The contributors guide uses syntax tests, but we cannot
```
Comment (ekpyron) [inline test/cmdlineTests.sh:184]: Ah, here we could use ``isoltest``, once it accepts files/directories as arguments.

**Review by ekpyron — COMMENTED** [review]
Looks fine apart from the type.

**Comment by chriseth** [comment]
Updated.

**Review by ekpyron — APPROVED** [review]

**Pull request merged**